Cherry-pick #8278 to 6.x: Allow TCP helper to support delimiters

jsoriano opened the pull request:
Cherry-pick of PR #8278 to 6.x branch. Original message: 

There are TCP payloads coming in from graphite carbon forwarders into metricbeat that have many metrics separated by '\n`. What happens is that only the first event is parsed and the rest are dropped.

This PR attempts to allow TCP helper to accept a delimiter character which can be used to split the bytes coming in to the socket and create events for each one of them.

ycombinator commented on metricbeat/helper/server/tcp/tcp.go:
```diff
@@ -96,22 +98,45 @@ func (g *TcpServer) watchMetrics() {
-		}()
 
-		length, err := conn.Read(buffer)
+		go g.handle(conn)
```
In the original PR, this statement is wrapped in an `if conn != nil`. Is that needed here too?

jsoriano commented:
@ycombinator thanks for reviewing this! And sorry for the confusion, I did a follow up PR (#8295)  to the original one with some extra details, this one backports both PRs, this is why you see these differences.

In the follow up PR I modified the changelog entry to be more explicit on the use case it supports. I also removed the deferred function enclosing `conn.Close()` (as it can be called directly), and moved the check for the nil connection. I consider that in the caller we shouldn't expect it to be nil as we have just checked that `Accept()` returned no error. In `handle()` it can still make sense defensively, in case it is called from somewhere else with a nil connection.
